### PR TITLE
SI-6183 don't crash on type error in outer test

### PR DESCRIPTION
**Pull request by adriaanm**
defend against ill-typed code reaching the outer-test codegen logic
since we already drop outer tests when in trouble, this fits nicely

there's no test case because the report was hard to minimize
since it was a crash on erroneous code, a backstop against a crash should suffice

review by @hubertp

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -1043,8 +1043,9 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
 
         def outerTest(testedBinder: Symbol, expectedTp: Type): Tree = {
           val expectedOuter = expectedTp.prefix match {
-            case ThisType(clazz)  => THIS(clazz)
-            case pre              => REF(pre.prefix, pre.termSymbol)
+            case ThisType(clazz)      => THIS(clazz)
+            case pre if pre != NoType => REF(pre.prefix, pre.termSymbol)
+            case _ => TRUE_typed // fallback for SI-6183
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1048]: Shouldn't this emit one or both of a log message and an unchecked warning? Or if "true" is the known right answer, this isn't exactly a fallback?

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1048]: I could emit a warning, but the only way `pre == NoType` I can think of would be for ill-typed code, so they'll get an error about that anyway.

Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1048]: Sounds ok

**Comment by hubertp** [comment]
lgtm

**Pull request merged**